umbraco/Umbraco-CMS
# Fix Sqlite database locking issue

## Issue
- When saving 2 document types (with templates) at the same time, the SqlLite database would totally lock up, and thus the whole application would be unusable.
This happens because when we save a document type, it needs to also create a template file on disk, thus it calls the FileService `CreateTemplateForContentType`, this file service would create a scope, get a read & write lock, and then call the `TemplateRepository`, that then would insert the template both in the database, but also create the file on disk.
All while the other thread would do the exact same steps as above, which would then result in the complete locking of the database. Why this results in the complete locking of the database I'm not sure, but it seems like both are taking writelocks, locking the database, but now its locked so neither can write.

Here is me replicating the issue:
![SqliteFreeze](https://user-images.githubusercontent.com/70372949/196933404-8c38309f-fa80-4e41-8ea4-3da5f53528f9.gif)

## Notes
- I have implemented a lock, that prevents simultaneous creation of these scopes in the `FileService`, which prevents the issue.
- This could also be solved by changing the `IsolationLevel` for the scope, so it would be `using ICoreScope scope = ScopeProvider.CreateCoreScope(IsolationLevel.UnCommited);` but this changed how strictly the database handles locking.


## How to test
- Run a fresh umbraco install in debug
- Set a breakpoint in `ContentTypeController` in the `PostSave` action on line 319.
![image](https://user-images.githubusercontent.com/70372949/196931329-c4d1b2c7-bfb3-4f50-8bd0-e2a762173cd5.png)
- Open backoffice in 2 tabs and create a document type in each (do not press save yet)
- When both is ready, save the first (it should hit the breakpoint), when the breakpoint is hit, save the other.
- Now both should be caught on the breakpoint which ensures they will run simultaneously.
- Remove the breakpoint and run the application (F5), this should no longer result in the locking of the database, and thus both document types should be saved 👍
